<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>gget enrichr - gget</title>


        <!-- Custom HTML head -->
        <!-- Google tag (gtag.js) -->
        <script async src="https://www.googletagmanager.com/gtag/js?id=G-ZM4N3XBEW7"></script>
        <script>
          window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());
        
          gtag('config', 'G-ZM4N3XBEW7');
        </script>

        <meta name="description" content="gget enables efficient querying of genomic reference databases">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="shortcut icon" href="../favicon.png">
        <link rel="stylesheet" href="../css/variables.css">
        <link rel="stylesheet" href="../css/general.css">
        <link rel="stylesheet" href="../css/chrome.css">
        <link rel="stylesheet" href="../css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../highlight.css">
        <link rel="stylesheet" href="../tomorrow-night.css">
        <link rel="stylesheet" href="../ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../en/introduction.html">Introduction</a></li><li class="chapter-item expanded affix "><a href="../en/updates.html">What's New?</a></li><li class="chapter-item expanded affix "><a href="../en/dependents.html">Dependents & News</a></li><li class="chapter-item expanded affix "><li class="part-title">User Guide</li><li class="chapter-item expanded "><a href="../en/installation.html">Installation</a></li><li class="chapter-item expanded "><a href="../en/quick_start_guide.html">Quick Start Guide</a></li><li class="chapter-item expanded affix "><li class="part-title">Manual</li><li class="chapter-item expanded "><a href="../en/alphafold.html">gget alphafold</a></li><li class="chapter-item expanded "><a href="../en/archs4.html">gget archs4</a></li><li class="chapter-item expanded "><a href="../en/bgee.html">gget bgee</a></li><li class="chapter-item expanded "><a href="../en/blast.html">gget blast</a></li><li class="chapter-item expanded "><a href="../en/blat.html">gget blat</a></li><li class="chapter-item expanded "><a href="../en/cbio.html">gget cbio</a></li><li class="chapter-item expanded "><a href="../en/cellxgene.html">gget cellxgene</a></li><li class="chapter-item expanded "><a href="../en/cosmic.html">gget cosmic</a></li><li class="chapter-item expanded "><a href="../en/diamond.html">gget diamond</a></li><li class="chapter-item expanded "><a href="../en/elm.html">gget elm</a></li><li class="chapter-item expanded "><a href="../en/enrichr.html">gget enrichr</a></li><li class="chapter-item expanded "><a href="../en/gpt.html">gget gpt</a></li><li class="chapter-item expanded "><a href="../en/info.html">gget info</a></li><li class="chapter-item expanded "><a href="../en/muscle.html">gget muscle</a></li><li class="chapter-item expanded "><a href="../en/mutate.html">gget mutate</a></li><li class="chapter-item expanded "><a href="../en/opentargets.html">gget opentargets</a></li><li class="chapter-item expanded "><a href="../en/pdb.html">gget pdb</a></li><li class="chapter-item expanded "><a href="../en/ref.html">gget ref</a></li><li class="chapter-item expanded "><a href="../en/search.html">gget search</a></li><li class="chapter-item expanded "><a href="../en/setup.html">gget setup</a></li><li class="chapter-item expanded "><a href="../en/seq.html">gget seq</a></li><li class="spacer"></li><li class="chapter-item expanded "><a href="../en/contributing.html">Contributing Guide</a></li><li class="chapter-item expanded "><a href="../en/code_of_conduct.html">Code of Conduct</a></li><li class="chapter-item expanded "><a href="../en/cite.html">Citation Guidelines</a></li><li class="spacer"></li><li class="chapter-item expanded affix "><li class="part-title">Español</li><li class="chapter-item expanded "><a href="../es/introduction.html">Introdución</a></li><li class="chapter-item expanded "><a href="../es/updates.html">¡Lo más reciente!</a></li><li class="chapter-item expanded "><a href="../es/dependents.html">Dependientes y Noticias</a></li><li class="chapter-item expanded affix "><li class="part-title">Guía del usario</li><li class="chapter-item expanded "><a href="../es/installation.html">Instalación</a></li><li class="chapter-item expanded "><a href="../es/quick_start_guide.html">Guía de inicio rápido</a></li><li class="chapter-item expanded affix "><li class="part-title">Manuál</li><li class="chapter-item expanded "><a href="../es/alphafold.html">gget alphafold</a></li><li class="chapter-item expanded "><a href="../es/archs4.html">gget archs4</a></li><li class="chapter-item expanded "><a href="../es/bgee.html">gget bgee</a></li><li class="chapter-item expanded "><a href="../es/blast.html">gget blast</a></li><li class="chapter-item expanded "><a href="../es/blat.html">gget blat</a></li><li class="chapter-item expanded "><a href="../es/cbio.html">gget cbio</a></li><li class="chapter-item expanded "><a href="../es/cellxgene.html">gget cellxgene</a></li><li class="chapter-item expanded "><a href="../es/cosmic.html">gget cosmic</a></li><li class="chapter-item expanded "><a href="../es/diamond.html">gget diamond</a></li><li class="chapter-item expanded "><a href="../es/elm.html">gget elm</a></li><li class="chapter-item expanded "><a href="../es/enrichr.html" class="active">gget enrichr</a></li><li class="chapter-item expanded "><a href="../es/gpt.html">gget gpt</a></li><li class="chapter-item expanded "><a href="../es/info.html">gget info</a></li><li class="chapter-item expanded "><a href="../es/muscle.html">gget muscle</a></li><li class="chapter-item expanded "><a href="../es/mutate.html">gget mutate</a></li><li class="chapter-item expanded "><a href="../es/opentargets.html">gget opentargets</a></li><li class="chapter-item expanded "><a href="../es/pdb.html">gget pdb</a></li><li class="chapter-item expanded "><a href="../es/ref.html">gget ref</a></li><li class="chapter-item expanded "><a href="../es/search.html">gget search</a></li><li class="chapter-item expanded "><a href="../es/setup.html">gget setup</a></li><li class="chapter-item expanded "><a href="../es/seq.html">gget seq</a></li><li class="spacer"></li><li class="chapter-item expanded "><a href="../es/contributing.html">Guía de contribución</a></li><li class="chapter-item expanded "><a href="../es/code_of_conduct.html">Codigo de conducto</a></li><li class="chapter-item expanded "><a href="../es/cite.html">Cómo citar</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">gget</h1>

                    <div class="right-buttons">
                        <a href="../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <blockquote>
<p>Parámetros de Python són iguales a los parámetros largos (<code>--parámetro</code>) de Terminal, si no especificado de otra manera. Las banderas son parámetros de verdadero o falso (True/False) en Python. El manuál para cualquier modulo de gget se puede llamar desde la Terminal con la bandera <code>-h</code> <code>--help</code>.</p>
</blockquote>
<h1 id="gget-enrichr-"><a class="header" href="#gget-enrichr-">gget enrichr 💰</a></h1>
<p>Realice un análisis de enriquecimiento de una lista de genes utilizando <a href="https://maayanlab.cloud/Enrichr/">Enrichr</a>.<br />
Produce: Resultados en formato JSON (Terminal) o Dataframe/CSV (Python).</p>
<p><strong>Parámetro posicional</strong><br />
<code>genes</code><br />
Lista de nombres cortos (símbolos) de los genes de interés para realizar el análisis de enriquecimiento, p. PHF14 RBM3 MSL1 PHF21A.<br />
Alternativamente: use la bandera <code>--ensembl</code> para ingresar IDs tipo Ensembl, p. ENSG00000106443 ENSG00000102317 ENSG00000188895.</p>
<p><strong>Otros parámetros requeridos</strong><br />
<code>-db</code> <code>--database</code><br />
Base de datos que será utilizada como referencia para el análisis de enriquecimiento.<br />
Admite cualquier base de datos enumerada <a href="https://maayanlab.cloud/Enrichr/#libraries">aquí</a> o uno de los siguientes accesos directos:<br />
'pathway'       (KEGG_2021_Human)<br />
'transcription'     (ChEA_2016)<br />
'ontology'      (GO_Biological_Process_2021)<br />
'diseases_drugs'   (GWAS_Catalog_2019)<br />
'celltypes'      (PanglaoDB_Augmented_2021)<br />
'kinase_interactions'   (KEA_2015)</p>
<p><strong>Parámetros opcionales</strong><br />
<code>-s</code> <code>--species</code><br />
Especies a utilizar como referencia para el análisis de enriquecimiento. (Por defecto: human)<br />
Opciones:</p>
<div class="table-wrapper"><table><thead><tr><th>Species</th><th>Database list</th></tr></thead><tbody>
<tr><td><code>human</code></td><td><a href="https://maayanlab.cloud/Enrichr/#libraries">Enrichr</a></td></tr>
<tr><td><code>mouse</code></td><td><a href="https://maayanlab.cloud/Enrichr/#libraries">Equivalente al humano</a></td></tr>
<tr><td><code>fly</code></td><td><a href="https://maayanlab.cloud/FlyEnrichr/#stats">FlyEnrichr</a></td></tr>
<tr><td><code>yeast</code></td><td><a href="https://maayanlab.cloud/YeastEnrichr/#stats">YeastEnrichr</a></td></tr>
<tr><td><code>worm</code></td><td><a href="https://maayanlab.cloud/WormEnrichr/#stats">WormEnrichr</a></td></tr>
<tr><td><code>fish</code></td><td><a href="https://maayanlab.cloud/FishEnrichr/#stats">FishEnrichr</a></td></tr>
</tbody></table>
</div>
<p><code>-bkg_l</code> <code>--background_list</code><br />
Lista de nombres cortos (símbolos) de genes de 'background' (de fondo/control), p. NSUN3 POLRMT NLRX1.<br />
Alternativamente: usa la bandera <code>--ensembl_background</code> para ingresar IDs tipo Ensembl.</p>
<p><code>-o</code> <code>--out</code><br />
Ruta al archivo en el que se guardarán los resultados, p. ruta/al/directorio/resultados.csv (o .json). Por defecto: salida estándar (STDOUT).<br />
Para Python, usa <code>save=True</code> para guardar los resultados en el directorio de trabajo actual.</p>
<p><code>-ko</code> <code>--kegg_out</code><br />
Ruta al archivo png en el que se guardará la imágen de la vía de señalización celular KEGG, p. ej. ruta/al/directorio/KEGG.png. (Por defecto: None)</p>
<p><code>-kr</code> <code>--kegg_rank</code><br />
Rango de la ruta KEGG que se va a trazar. (Por defecto: 1)</p>
<p><code>figsize</code><br />
Solo para Python. (ancho, alto) de la visualización en pulgadas. (Por defecto: (10,10))</p>
<p><code>ax</code><br />
Solo para Python. Ingresa un objeto de ejes matplotlib para personalizar la visualización.(Por defecto: None)</p>
<p><strong>Banderas</strong><br />
<code>-e</code> <code>--ensembl</code><br />
Usa esta bandera si <code>genes</code> se ingresa como una lista de IDs tipo Ensembl.</p>
<p><code>-e_b</code> <code>--ensembl_bkg</code><br />
Usa esta bandera si <code>background_list</code> se ingresa como una lista de IDs tipo Ensembl.</p>
<p><code>-bkg</code> <code>--background</code><br />
Use un conjunto de 20,625 genes 'background' 
listados <a href="https://github.com/pachterlab/gget/blob/main/gget/constants/enrichr_bkg_genes.txt">aquí</a>.</p>
<p><code>-csv</code> <code>--csv</code><br />
Solo para Terminal. Produce los resultados en formato CSV.<br />
Para Python, usa <code>json=True</code> produce los resultados en formato JSON.</p>
<p><code>-q</code> <code>--quiet</code><br />
Solo para Terminal. Impide la información de progreso de ser exhibida durante la ejecución del programa.<br />
Para Python, usa <code>verbose=False</code> para imipidir la información de progreso de ser exhibida durante la ejecución del programa.</p>
<p><code>plot</code><br />
Solo para Python. <code>plot=True</code> provée la visualización de los primeros 15 resultados (por defecto: False).</p>
<h3 id="ejemplo"><a class="header" href="#ejemplo">Ejemplo</a></h3>
<pre><code class="language-bash">gget enrichr -db ontology ACE2 AGT AGTR1
</code></pre>
<pre><code class="language-python"># Python
gget.enrichr([&quot;ACE2&quot;, &quot;AGT&quot;, &quot;AGTR1&quot;], database=&quot;ontology&quot;, plot=True)
</code></pre>
<p>→ Produce vías/funciones celulares relacionadas con los genes ACE2, AGT y AGTR1 de la base de datos <em>GO Biological Process 2021</em>. En Python, <code>plot=True</code> provee la visualización de resultados:</p>
<p><img src="https://github.com/pachterlab/gget/blob/main/figures/gget_enrichr_results.png?raw=true" alt="alt text" /></p>
<p><br/><br/></p>
<p><strong>Use <code>gget enrichr</code> con una lista de genes 'background':</strong></p>
<pre><code class="language-bash"># Aquí, primero ingresamos los genes de interés (parámetro posicional 'genes'), para que no se agreguen a la lista de genes 'background' detrás del parámetro '-bkgr_l'
gget enrichr \
	PHF14 RBM3 MSL1 PHF21A ARL10 INSR JADE2 P2RX7 LINC00662 CCDC101 PPM1B KANSL1L CRYZL1 ANAPC16 TMCC1 CDH8 RBM11 CNPY2 HSPA1L CUL2 PLBD2 LARP7 TECPR2 ZNF302 CUX1 MOB2 CYTH2 SEC22C EIF4E3 ROBO2 ADAMTS9-AS2 CXXC1 LINC01314 ATF7 ATP5F1 \
	-db ChEA_2022 \
	-bkg_l NSUN3 POLRMT NLRX1 SFXN5 ZC3H12C SLC25A39 ARSG DEFB29 PCMTD2 ACAA1A LRRC1 2810432D09RIK SEPHS2 SAC3D1 TMLHE LOC623451 TSR2 PLEKHA7 GYS2 ARHGEF12 HIBCH LYRM2 ZBTB44 ENTPD5 RAB11FIP2 LIPT1 INTU ANXA13 KLF12 SAT2 GAL3ST2 VAMP8 FKBPL AQP11 TRAP1 PMPCB TM7SF3 RBM39 BRI3 KDR ZFP748 NAP1L1 DHRS1 LRRC56 WDR20A STXBP2 KLF1 UFC1 CCDC16 9230114K14RIK RWDD3 2610528K11RIK ACO1 CABLES1 LOC100047214 YARS2 LYPLA1 KALRN GYK ZFP787 ZFP655 RABEPK ZFP650 4732466D17RIK EXOSC4 WDR42A GPHN 2610528J11RIK 1110003E01RIK MDH1 1200014M14RIK AW209491 MUT 1700123L14RIK 2610036D13RIK PHF14 RBM3 MSL1 PHF21A ARL10 INSR JADE2 P2RX7 LINC00662 CCDC101 PPM1B KANSL1L CRYZL1 ANAPC16 TMCC1 CDH8 RBM11 CNPY2 HSPA1L CUL2 PLBD2 LARP7 TECPR2 ZNF302 CUX1 MOB2 CYTH2 SEC22C EIF4E3 ROBO2 ADAMTS9-AS2 CXXC1 LINC01314 ATF7 ATP5F1COX15 TMEM30A NSMCE4A TM2D2 RHBDD3 ATXN2 NFS1 3110001I20RIK BC038156 C330002I19RIK ZFYVE20 POLI TOMM70A LOC100047782 2410012H22RIK RILP A230062G08RIK PTTG1IP RAB1 AFAP1L1 LYRM5 2310026E23RIK SLC7A6OS MAT2B 4932438A13RIK LRRC8A SMO NUPL2
</code></pre>
<pre><code class="language-python"># Python
gget.enrichr(
	genes = [
		&quot;PHF14&quot;, &quot;RBM3&quot;, &quot;MSL1&quot;, &quot;PHF21A&quot;, &quot;ARL10&quot;, &quot;INSR&quot;, &quot;JADE2&quot;, &quot;P2RX7&quot;,
		&quot;LINC00662&quot;, &quot;CCDC101&quot;, &quot;PPM1B&quot;, &quot;KANSL1L&quot;, &quot;CRYZL1&quot;, &quot;ANAPC16&quot;, &quot;TMCC1&quot;,
		&quot;CDH8&quot;, &quot;RBM11&quot;, &quot;CNPY2&quot;, &quot;HSPA1L&quot;, &quot;CUL2&quot;, &quot;PLBD2&quot;, &quot;LARP7&quot;, &quot;TECPR2&quot;, 
		&quot;ZNF302&quot;, &quot;CUX1&quot;, &quot;MOB2&quot;, &quot;CYTH2&quot;, &quot;SEC22C&quot;, &quot;EIF4E3&quot;, &quot;ROBO2&quot;,
		&quot;ADAMTS9-AS2&quot;, &quot;CXXC1&quot;, &quot;LINC01314&quot;, &quot;ATF7&quot;, &quot;ATP5F1&quot;
	], 
	database = &quot;ChEA_2022&quot;,
	background_list = [
		&quot;NSUN3&quot;,&quot;POLRMT&quot;,&quot;NLRX1&quot;,&quot;SFXN5&quot;,&quot;ZC3H12C&quot;,&quot;SLC25A39&quot;,&quot;ARSG&quot;,
		&quot;DEFB29&quot;,&quot;PCMTD2&quot;,&quot;ACAA1A&quot;,&quot;LRRC1&quot;,&quot;2810432D09RIK&quot;,&quot;SEPHS2&quot;,
		&quot;SAC3D1&quot;,&quot;TMLHE&quot;,&quot;LOC623451&quot;,&quot;TSR2&quot;,&quot;PLEKHA7&quot;,&quot;GYS2&quot;,&quot;ARHGEF12&quot;,
		&quot;HIBCH&quot;,&quot;LYRM2&quot;,&quot;ZBTB44&quot;,&quot;ENTPD5&quot;,&quot;RAB11FIP2&quot;,&quot;LIPT1&quot;,
		&quot;INTU&quot;,&quot;ANXA13&quot;,&quot;KLF12&quot;,&quot;SAT2&quot;,&quot;GAL3ST2&quot;,&quot;VAMP8&quot;,&quot;FKBPL&quot;,
		&quot;AQP11&quot;,&quot;TRAP1&quot;,&quot;PMPCB&quot;,&quot;TM7SF3&quot;,&quot;RBM39&quot;,&quot;BRI3&quot;,&quot;KDR&quot;,&quot;ZFP748&quot;,
		&quot;NAP1L1&quot;,&quot;DHRS1&quot;,&quot;LRRC56&quot;,&quot;WDR20A&quot;,&quot;STXBP2&quot;,&quot;KLF1&quot;,&quot;UFC1&quot;,
		&quot;CCDC16&quot;,&quot;9230114K14RIK&quot;,&quot;RWDD3&quot;,&quot;2610528K11RIK&quot;,&quot;ACO1&quot;,
		&quot;CABLES1&quot;, &quot;LOC100047214&quot;,&quot;YARS2&quot;,&quot;LYPLA1&quot;,&quot;KALRN&quot;,&quot;GYK&quot;,
		&quot;ZFP787&quot;,&quot;ZFP655&quot;,&quot;RABEPK&quot;,&quot;ZFP650&quot;,&quot;4732466D17RIK&quot;,&quot;EXOSC4&quot;,
		&quot;WDR42A&quot;,&quot;GPHN&quot;,&quot;2610528J11RIK&quot;,&quot;1110003E01RIK&quot;,&quot;MDH1&quot;,&quot;1200014M14RIK&quot;,
		&quot;AW209491&quot;,&quot;MUT&quot;,&quot;1700123L14RIK&quot;,&quot;2610036D13RIK&quot;,
		&quot;PHF14&quot;, &quot;RBM3&quot;, &quot;MSL1&quot;, &quot;PHF21A&quot;, &quot;ARL10&quot;, &quot;INSR&quot;, &quot;JADE2&quot;, 
		&quot;P2RX7&quot;, &quot;LINC00662&quot;, &quot;CCDC101&quot;, &quot;PPM1B&quot;, &quot;KANSL1L&quot;, &quot;CRYZL1&quot;, 
		&quot;ANAPC16&quot;, &quot;TMCC1&quot;,&quot;CDH8&quot;, &quot;RBM11&quot;, &quot;CNPY2&quot;, &quot;HSPA1L&quot;, &quot;CUL2&quot;, 
		&quot;PLBD2&quot;, &quot;LARP7&quot;, &quot;TECPR2&quot;, &quot;ZNF302&quot;, &quot;CUX1&quot;, &quot;MOB2&quot;, &quot;CYTH2&quot;, 
		&quot;SEC22C&quot;, &quot;EIF4E3&quot;, &quot;ROBO2&quot;, &quot;ADAMTS9-AS2&quot;, &quot;CXXC1&quot;, &quot;LINC01314&quot;, &quot;ATF7&quot;, 
		&quot;ATP5F1&quot;&quot;COX15&quot;,&quot;TMEM30A&quot;,&quot;NSMCE4A&quot;,&quot;TM2D2&quot;,&quot;RHBDD3&quot;,&quot;ATXN2&quot;,&quot;NFS1&quot;,
		&quot;3110001I20RIK&quot;,&quot;BC038156&quot;,&quot;C330002I19RIK&quot;,&quot;ZFYVE20&quot;,&quot;POLI&quot;,&quot;TOMM70A&quot;,
		&quot;LOC100047782&quot;,&quot;2410012H22RIK&quot;,&quot;RILP&quot;,&quot;A230062G08RIK&quot;,
		&quot;PTTG1IP&quot;,&quot;RAB1&quot;,&quot;AFAP1L1&quot;, &quot;LYRM5&quot;,&quot;2310026E23RIK&quot;,
		&quot;SLC7A6OS&quot;,&quot;MAT2B&quot;,&quot;4932438A13RIK&quot;,&quot;LRRC8A&quot;,&quot;SMO&quot;,&quot;NUPL2&quot;
	],
	plot=True
)
</code></pre>
<p>→ Provée factores de transcripción relacionados a los genes de interés y controlados con la lista de genes background de la base de datos <em>ChEA 2022</em>. En Python, <code>plot=True</code> permite la visualización de resultados:</p>
<p><img src="https://github.com/pachterlab/gget/blob/main/figures/gget_enrichr_results_2.png?raw=true" alt="alt text" /></p>
<p><br/><br/></p>
<p><strong>Genere una imagen de la vía de señalización de células KEGG con los genes del análisis de enriquecimiento resaltados:</strong><br />
Esta función está disponible gracias a un <a href="https://github.com/pachterlab/gget/pull/106">PR</a> de <a href="https://github.com/noriakis">Noriaki Sato</a>.</p>
<pre><code class="language-bash">gget enrichr -db pathway --kegg_out kegg.png --kegg_rank 1 ZBP1 IRF3 RIPK1
</code></pre>
<pre><code class="language-python"># Python
gget.enrichr([&quot;ZBP1&quot;, &quot;IRF3&quot;, &quot;RIPK1&quot;], database=&quot;pathway&quot;, kegg_out=&quot;kegg.png&quot;, kegg_rank=1)
</code></pre>
<p>→ Además de los resultados estándar <code>gget enrichr</code>, el argumento <code>kegg_out</code> guarda una imagen con los genes del análisis de enriquecimiento resaltados guardado como kegg.png:</p>
<p><img src="https://github.com/pachterlab/gget/assets/56094636/b0aa5a64-69d0-4a6a-85db-3e7baf9cb2a4" alt="kegg" /></p>
<p><br/><br/></p>
<p>El siguiente ejemplo fue enviado por <a href="https://github.com/DylanLawless">Dylan Lawless</a> a través de un <a href="https://github.com/pachterlab/gget/pull/54">PR</a> (con ajustes de <a href="https://github.com/lauraluebbert">Laura Luebbert</a>):<br />
<strong>Use <code>gget enrichr</code> en R y cree unq visualización similar usando <a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a>.</strong><br />
TENGA EN CUENTA el cambio de ejes en comparación con la visualización en Python. </p>
<pre><code class="language-r">system(&quot;pip install gget&quot;)
install.packages(&quot;reticulate&quot;)
library(reticulate)
gget &lt;- import(&quot;gget&quot;)

# Perform enrichment analysis on a list of genes
df &lt;- gget$enrichr(list(&quot;ACE2&quot;, &quot;AGT&quot;, &quot;AGTR1&quot;), database = &quot;ontology&quot;)

# Count number of overlapping genes
df$overlapping_genes_count &lt;- lapply(df$overlapping_genes, length) |&gt; as.numeric()

# Only keep the top 15 results
df &lt;- df[1:15, ]

# Plot
library(ggplot2)

df |&gt;
	ggplot() +
	geom_bar(aes(
		x = -log10(adj_p_val),
		y = reorder(path_name, -adj_p_val)
	),
	stat = &quot;identity&quot;,
  	fill = &quot;lightgrey&quot;,
  	width = 0.5,
	color = &quot;black&quot;) +
	geom_text(
		aes(
			y = path_name,
			x = (-log10(adj_p_val)),
			label = overlapping_genes_count
		),
		nudge_x = 0.75,
		show.legend = NA,
		color = &quot;red&quot;
	) +
  	geom_text(
		aes(
			y = Inf,
			x = Inf,
      			hjust = 1,
      			vjust = 1,
			label = &quot;# of overlapping genes&quot;
		),
		show.legend = NA,
		color = &quot;red&quot;
	) +
	geom_vline(linetype = &quot;dotted&quot;, linewidth = 1, xintercept = -log10(0.05)) +
	ylab(&quot;Pathway name&quot;) +
	xlab(&quot;-log10(adjusted P value)&quot;)
</code></pre>
<h4 id="más-ejemplos"><a class="header" href="#más-ejemplos"><a href="https://github.com/pachterlab/gget_examples">Más ejemplos</a></a></h4>
<h1 id="citar"><a class="header" href="#citar">Citar</a></h1>
<p>Si utiliza <code>gget enrichr</code> en una publicación, favor de citar los siguientes artículos:</p>
<ul>
<li>
<p>Luebbert, L., &amp; Pachter, L. (2023). Efficient querying of genomic reference databases with gget. Bioinformatics. <a href="https://doi.org/10.1093/bioinformatics/btac836">https://doi.org/10.1093/bioinformatics/btac836</a></p>
</li>
<li>
<p>Chen EY, Tan CM, Kou Y, Duan Q, Wang Z, Meirelles GV, Clark NR, Ma'ayan A. Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC Bioinformatics. 2013; 128(14). <a href="https://doi.org/10.1186/1471-2105-14-128">https://doi.org/10.1186/1471-2105-14-128 </a> </p>
</li>
<li>
<p>Kuleshov MV, Jones MR, Rouillard AD, Fernandez NF, Duan Q, Wang Z, Koplev S, Jenkins SL, Jagodnik KM, Lachmann A, McDermott MG, Monteiro CD, Gundersen GW, Ma'ayan A. Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic Acids Research. 2016; gkw377. doi: <a href="https://doi.org/10.1093/nar/gkw377">10.1093/nar/gkw377</a> </p>
</li>
<li>
<p>Xie Z, Bailey A, Kuleshov MV, Clarke DJB., Evangelista JE, Jenkins SL, Lachmann A, Wojciechowicz ML, Kropiwnicki E, Jagodnik KM, Jeon M, &amp; Ma’ayan A. Gene set knowledge discovery with Enrichr. Current Protocols, 1, e90. 2021. doi: <a href="https://doi.org/10.1002/cpz1.90">10.1002/cpz1.90</a>.</p>
</li>
</ul>
<p>Si trabaja con conjuntos de datos no humanos/ratón, cite también:</p>
<ul>
<li>Kuleshov MV, Diaz JEL, Flamholz ZN, Keenan AB, Lachmann A, Wojciechowicz ML, Cagan RL, Ma'ayan A. modEnrichr: a suite of gene set enrichment analysis tools for model organisms. Nucleic Acids Res. 2019 Jul 2;47(W1):W183-W190. doi: <a href="https://doi.org/10.1093/nar/gkz347">10.1093/nar/gkz347</a>. PMID: 31069376; PMCID: PMC6602483.</li>
</ul>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../es/elm.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="../es/gpt.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../es/elm.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="../es/gpt.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="../elasticlunr.min.js"></script>
        <script src="../mark.min.js"></script>
        <script src="../searcher.js"></script>

        <script src="../clipboard.min.js"></script>
        <script src="../highlight.js"></script>
        <script src="../book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
